### PR TITLE
Tree walk 

### LINKED ISSUE #52: Add `walk` generator to Tree.
It would be a nice convenience to walk a Tree precisely like `scandir.walk` like:

```python
import datreant.core as dtr

for root, dirs, files in dtr.Tree('mytree').walk():
    # do stuff
```

This should be pretty straightforward.

### PR DESCRIPTION
**Pull request by jdetle**
Fixes #52.
Right now I have strange bug in tests where tree.walk (which is just a simple wrapper for scandir.walk() ) adds a forward slash to the root paths generated by scandir. I have no idea why, I can easily delete the last element from the string but I wanted your input in case that this is a sign I am doing something wrong.

### CODE REVIEW

File: src/datreant/core/trees.py [hunk 1]
```diff
@@ -467,3 +467,17 @@ def make(self):
         self.makedirs()
 
         return self
+
+    def walk(self, topdown=True, onerror=None, followlinks=False):
+        """Generate the filenames in the tree by walking the tree either
+           top-down or bottom-up.  For each directory in the tree rooted at
+           directory top (including top itself), it yields a 3-tuple
+           (dirpath, dirnames, filenames).
+
```
Comment (dotsdl) [inline src/datreant/core/trees.py]: Please make this a ``numpy`` style docstring. See the surrounding docstrings for other methods for the style.

File: src/datreant/core/trees.py [hunk 2]
```diff
@@ -467,3 +467,17 @@ def make(self):
...
+           (Documentation taken from os.walk)
+        """
+        if not self.exists:
+            raise OSError("Tree doesn't exist in the filesystem")
+        else:
+            return scandir.walk(self.abspath, topdown=topdown, onerror=onerror,
+                                followlinks=False)
```
Comment (dotsdl) [inline src/datreant/core/trees.py]: Should this be `followlinks=True`?

Comment (jdetle) [inline src/datreant/core/trees.py]: yes, sorry!

**Comment by dotsdl** [comment]
> Right now I have strange bug in tests where tree.walk (which is just a simple wrapper for scandir.walk() ) adds a forward slash to the root paths generated by scandir. I have no idea why, I can easily delete the last element from the string but I wanted your input in case that this is a sign I am doing something wrong.

@jdetle I'm not seeing the same issue that you describe. For me, doing:

```python
import datreant.core as dtr

s = dtr.Treant('sprout')

for root, dirs, files in s.walk():
    print(root)
    print(dirs)
    print(files)
    print('------')

```
shows the `root` paths *lacking* a forward slash at the end, whereas it's customary in `datreant` to make directory paths clearly distinctive by adding a slash at the end. It's not a big deal, here, though, since anyone working with paths should use things like `os.path.join` to join dirs and filenames for full paths.

**Comment by dotsdl** [comment]
I did just have an idea that might make this more useful: if instead of yielding raw strings and lists of them for `root`, `dirs`, and `files`, perhaps we should use our own filesystem objects and yield `Tree` objects, and `View`s of `Tree`s and `Leaf`s (for `dirs` and `files`, respectively). 

This would require wrapping the generator somehow (I think a `decorator` would work?) that turns the yielded components into the kinds of things we want instead. Wanna give it a try? This might come at an appreciable performance cost, but I would like to know to what degree. Probably not much compared to querying the filesystem.

**Comment by jdetle** [comment]
Sure, we could also just go in and borrow the scandir code and apply it to the objects you just said.

**Comment by dotsdl** [comment]
> Sure, we could also just go in and borrow the scandir code and apply it to the objects you just said, (if i understand correctly)

@jdetle we should avoid copying-and-pasting the `scandir.walk` code directly. I think the easiest option is for `Tree.walk` to return a generator as it already does, but this generator should be our own that internally yields results from the generator returned by `scandir.walk`. This wrapping generator just repackages those results in terms of our favorite filesystem objects.

Does that make sense? I'm traveling at the moment, but I could give a rough code example if you like.

**Comment by jdetle** [comment]
Yea I think so! This seems like it would make sense in terms of what a user would want out of a walk.

**Comment by jdetle** [comment]
Hey, sorry for being a drive-by committer! I'll try to get this done this done this weekend.

**Comment by jdetle** [comment]
@dotsdl, is this what you meant?

**Comment by dotsdl** [comment]
Hi @jdetle, I took your branch and made some adjustments. Check out [tree_walk](https://github.com/datreant/datreant.core/tree/tree_walk), and tell me what you think. You're welcome to merge this into your own branch and go from there.

**Comment by dotsdl** [comment]
Haven't had a look at your tests yet. Basically I would compare the relative paths given by each `View` in the iteration to what you should get if you were using `scandir.walk`. From my quick glance it looks like you're on that track.

**Comment by jdetle** [comment]
Ah! I think I misunderstood some of the core machinery of how trees and leaves work, thanks. Will do, (I'm probably going to work through lunch because its so oppressively hot here.)

File: src/datreant/core/trees.py [hunk 3]
```diff
@@ -525,9 +525,58 @@ def sync(self, other, mode='upload', compress=True, checksum=True,
         # Here we do some massaging before passing to the rsync function
         return rsync(source, dest, compress=compress, backup=backup,
                      dry=dry, include=include, checksum=checksum,
+<<<<<<< HEAD
```
Comment (jbarnoud) [inline src/datreant/core/trees.py]: Caution! You have unresolved conflicts here.

File: src/datreant/core/trees.py [hunk 4]
```diff
@@ -525,9 +525,58 @@ def sync(self, other, mode='upload', compress=True, checksum=True,
...
+        generator
+            Wrapped `scandir.walk()` generator yielding `datreant` objects
+
+        """
+        from .collections import View
+
+        if not self.exists:
+            raise OSError("Tree doesn't exist in the filesystem")
+
+<<<<<<< HEAD
```
Comment (jbarnoud) [inline src/datreant/core/trees.py]: Unresolved conflicts here too.

Comment (jdetle) [inline src/datreant/core/trees.py]: Too much coffee!

**Comment by dotsdl** [comment]
Closing since we are now working from #86.

**Pull request closed without merging**